src/en/charms-constraints.md

```diff
@@ -1,187 +1,233 @@
-constraints (`juju add-machine`) and then specify that machine when deploying 
-applications ([see the documentation on `juju deploy`](./charms-deploying.html)).
+The idealized use case is that of stipulating a constraint when deploying an
+application and the backing cloud providing a machine with those exact
+resources. In the majority of cases, however, default constraints may have been
+set (at various levels) and the cloud is unable to supply those exact
+resources.
 
-## What constraints can be used?
+Several noteworthy constraint characteristics:
 
-There is a full list of the constraints used 
-[in the reference section](reference-constraints.html). Be aware that some of 
-these are specific to the type of cloud you are using. For example, one may
-understand an "instance-type" constraint, but another may not. 
+ - Some constraints are only supported by certain clouds.
+ - Changes to constraint defaults do not affect existing machines.
+ - Multiple constraints are logically AND'd (i.e. the machine must satisfy all
+   constraints).
 
-The most useful constraints for Juju in general are:
-  
-  - **mem** : This indicates the minimum number of megabytes of RAM that must 
-  be available to an application unit. An optional suffix of M/G/T/P indicates
-  the value is mega-/giga-/tera-/peta- bytes.
-
-  - **cores** :  How many cpu cores the host machine should have. This is a
-  crude indicator of system performance.
-    
-  - **spaces** : Target a particular network space, or avoid one. See 
-  [Network spaces][network-spaces].
-  
-  - **arch** : Short for 'architecture', indicates the processor type an
-  application must run on. One of amd64, arm, i386, arm64, or ppc64el.
-  
-With these you can make sure an application has the resources it needs to run 
-properly.
+## Clouds and constraints
 
-Constraints can be used with commands that support the '--constraints' option. 
-These are covered in more detail below, but there are some aspects of specifying
-constraints that are worth mentioning first.
+When the backing cloud is unable to precisely satisfy a constraint, the
+resulting system's resources will exceed the constraint-defined minimum.
+However, if the cloud cannot satisfy a constraint at all then an error will be
+emitted and a machine will not be provisioned.
 
-In the following examples, we will be using the `juju deploy` command, as this
-is the simplest and most frequent case for using constraints. So, to deploy the
-'mariadb' application to a machine with 4 gigabytes of memory or more:
-  
-```bash
-juju deploy mariadb --constraints mem=4G
-```
+## Constraint scopes, defaults, and precedence
 
-To further ensure that it also has at least 2 CPU cores:
-  
-```bash
-juju deploy mariadb --constraints "mem=4G cores=2"
-```
+Constraints can be applied to various levels or scopes. Defaults can be set on
+some of them, and in the case of overlapping configurations a precedence is
```

Comment:
this seems vague at this point. Also, the only things you can't apply 'defaults' to are things which make no sense - controller - there is only one. Single-machine, it depends what you mean by that. If you mean a machine generated by 'add-machine', there is a default for that. The only way I can thing you can make a single machine that doesn't inherit constraints if they are set is if you manually add it to a cloud. Am I missing something?

The point to my comment is, this information would be a lot easier to digest if there were less of it, so leaving out a section on 'defaults' and an additional list would be a bonus.

Reply:
AFAIK, you cannot specify a default for the addition of machines. You can set one for a model, though, which affects `add-machine`. This is analogous to setting a default for all models, which affects `add-model`, but there is a definite default that can be set for 'all-models' and a definite default that can be set for 'a model'.

As you know, I'm all for less text but default settings are a central part of constraints. The user should be told about them upfront.